airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/global_substream_cursor.py

```diff
@@ -0,0 +1,269 @@
+        does not have parent streams, this step will be skipped due to the default PartitionRouter implementation.
+
+        Args:
+            stream_state (StreamState): The state of the streams to be set. The format of the stream state should be:
+                {
+                    "state": {
+                        "last_updated": "2023-05-27T00:00:00Z"
+                    },
+                    "parent_state": {
+                        "parent_stream_name": {
+                            "last_updated": "2023-05-27T00:00:00Z"
+                        }
+                    },
+                    "lookback_window": 132
+                }
+        """
+        if not stream_state:
+            return
+
+        if "lookback_window" in stream_state:
+            self._lookback_window = stream_state["lookback_window"]
+            self._inject_lookback_into_stream_cursor(stream_state["lookback_window"])
+
+        self._stream_cursor.set_initial_state(stream_state["state"])
+
+        # Set parent state for partition routers based on parent streams
+        self._partition_router.set_initial_state(stream_state)
+
+    def _inject_lookback_into_stream_cursor(self, lookback_window: int) -> None:
+        """
+        Modifies the stream cursor's lookback window based on the duration of the previous sync.
+        This adjustment ensures the cursor is set to the minimal lookback window necessary for
+        avoiding missing data.
+
+        Parameters:
+            lookback_window (int): The lookback duration in seconds to be set, derived from
+                                   the previous sync.
+
+        Raises:
+            ValueError: If the cursor does not support dynamic lookback window adjustments.
+        """
+        if hasattr(self._stream_cursor, "set_runtime_lookback_window"):
+            self._stream_cursor.set_runtime_lookback_window(lookback_window)
+        else:
+            raise ValueError("The cursor class for Global Substream Cursor does not have a set_runtime_lookback_window method")
+
+    def observe(self, stream_slice: StreamSlice, record: Record) -> None:
+        self._stream_cursor.observe(StreamSlice(partition={}, cursor_slice=stream_slice.cursor_slice), record)
+
+    def close_slice(self, stream_slice: StreamSlice, *args: Any) -> None:
+        """
+        Close the current stream slice.
+
+        This method is called when a stream slice is completed. For the global parent cursor, we close the child cursor
+        only after reading all slices. This ensures that we do not miss any child records from a later parent record
+        if the child cursor is earlier than a record from the first parent record.
+
+        Args:
+            stream_slice (StreamSlice): The stream slice to be closed.
+            *args (Any): Additional arguments.
+        """
+        with self._lock:
+            self._slice_semaphore.acquire()
+            if self._all_slices_yielded and self._slice_semaphore._value == 0:
```

Comment:
I don't know much about the Python threading framework but I fear that this might not be thread safe. There are no lock on the `Semaphore._value` manipulations so it feels like we could end up in a case there `self._slice_semaphore._value` is not what we expect it to be. I think that in our case, it might be fine because we only check once everything is done, but this couple with the fact that we update `_all_slices_yielded ` before yielding the last slice makes me a bit worried. Do you know how we could prove this is really thread safe? Or should we just have our own integer value that we always access behind `with self._lock`?

Reply:
Only one thread can execute close_slice at a time due to the lock, ensuring thread safety. The `_all_slices_yielded` flag is set before yielding the last slice, guaranteeing that `close_slice` will always be called at least once with `_all_slices_yielded` set to `True`. When `_all_slices_yielded` is `True`, there will be exactly one instance where the semaphore value reaches 0, which will trigger the final `close_slice` operation. The semaphore is thread-safe, so there shouldn’t be any issues with counting, and the check for 0 is thread-safe and only done by one thread at a time.